src/bigint.rs

```diff
@@ -709,77 +921,88 @@ impl Integer for BigUint {
-            while m >= b {
-                let (d0, d_unit, b_unit) = div_estimate(&m, &b, n);
-                let mut d0 = d0;
-                let mut prod = &b * &d0;
-                while prod > m {
-                    // FIXME(#5992): assignment operator overloads
-                    // d0 -= &d_unit
-                    d0   = d0 - &d_unit;
-                    // FIXME(#5992): assignment operator overloads
-                    // prod -= &b_unit;
-                    prod = prod - &b_unit
-                }
-                if d0.is_zero() {
-                    n = 2;
+
+        fn div_mod_floor_inner(mut a: BigUint, b: BigUint) -> (BigUint, BigUint) {
+            /*
+             * The algorithm works by incrementally calculating "guesses", q0, for part of the
+             * remainder. Once we have any number q0 such that q0 * b <= a, we can set
+             *
+             *     q += q0
+             *     a -= q0 * b
+             *
+             * and then iterate until a < b. Then, (q, a) will be our desired quotient and
+             * remainder.
+             *
+             * q0, our guess, is calculated by dividing the last few digits of a by the last digit
+             * of b - this should give us a guess that is "close" to the actual quotient, but is
+             * possibly greater than the actual quotient. If q0 * b > a, we simply use iterated
+             * subtraction until we have a guess such that q8 & b <= a.
+             */
+
+            let bn = *b.data.last().unwrap();
+            let q_len = a.data.len() - b.data.len() + 1;
+            let mut q: BigUint = BigUint { data: Vec::with_capacity(q_len) };
+
+            q.data.extend(repeat(0).take(q_len));
+
+            /* We reuse the same temporary to avoid hitting the allocator in our inner loop: */
+            let mut tmp: BigUint = BigUint { data: Vec::with_capacity(2) };
+
+            for j in (0..q_len).rev() {
+                /*
+                 * When calculating our next guess q0, we don't need to consider the digits below j
+                 * + b.data.len() - 1: we're guessing digit j of the quotient (i.e. q0 << j) from
+                 * digit bn of the divisor (i.e. bn << (b.data.len() - 1) - so the product of those
+                 * two numbers will be zero in all digits up to (j + b.data.len() - 1).
+                 */
+                let offset = j + b.data.len() - 1;
+                if offset >= a.data.len() {
                     continue;
                 }
-                n = 1;
-                // FIXME(#5992): assignment operator overloads
-                // d += d0;
-                d = d + d0;
-                // FIXME(#5992): assignment operator overloads
-                // m -= prod;
-                m = m - prod;
+
+                /* just avoiding a heap allocation: */
+                let mut a0 = tmp;
+                a0.data.truncate(0);
+                a0.data.extend(&a.data[offset..]);
```

Comment:
This failed CI on Rust 1.0.  Later Rust has for convenience:

    impl<'a, T> Extend<&'a T> for Vec<T> where T: Copy + 'a

But for maintaining 1.0 compatibility, we'll have to explicitly call `a.data[offset..].iter().cloned()`.  This shouldn't make any difference in performance, just syntactic sugar.